### PR TITLE
Issue 7272: Add additional checkstyle checks

### LINKED ISSUE #7272: Add additional checkstyle checks
**Describe the bug**
We are running an outdated version of checkstyle. The newer version has many additional checks which we could enable.
Additionally it is possible to define our own checks. It would be good to add checks for:
* Sleeping in a unit test
* Static atomic references
* Break labels

**Proposed fix**
Update to the newest version of checkstyle and enable additional checks.

### PR DESCRIPTION
**Pull request by tkaitchuck**
**Change log description**  
* Enable many additional checkstyle rules

**Purpose of the change**  
Fix #7272

**What the code does**  
Enables checks to prohibit:
* Double brace initialization
* Use of the `volatile` keyword
* Use of break labels
* Use of Finalizers
* Interfaces with member variables
* Mutable exceptions
* Static AtomicReferences
* Sleeping in unit tests
* Empty catch blocks

Additionally add checks to prevent accidently:
* Forgetting to override equals when required
* Forgetting to write Javadocs in client / common

Finally it imposes limits on:
* Line length (200)
* Nesting depth
* Boolean expression complexity
* Method length
* Number of parameters

To avoid a large messy change with the exception of line length and a few minor fixes, existing classes/methods which would not comply with the new checkstyle rules are whitelisted in a `grandfathered` list. This exempts those specific instances from the checks they currently violate but prevents other code from following the same pattern.

### CODE REVIEW

File: checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -116,24 +181,26 @@
...
-      <property name="allowMissingThrowsTags" value="false"/>
+      <property name="accessModifiers" value="public, protected"/>
+      <property name="validateThrows" value="false"/>
       <property name="allowMissingParamTags" value="false"/>
       <property name="allowMissingReturnTag" value="true"/>
-      <property name="allowMissingJavadoc" value="true"/> <!--TODO: this should be enabled at one point. -->
-      <property name="ignoreMethodNamesRegex" value="^get.*$"/> <!--It would be nice if we could allow checking that the doc exists without also verifying the @returns is there also, but checkstyle does not allow that. -->
-      <property name="ignoreMethodNamesRegex" value="^has.*$"/> <!-- accessor for boolean e.g. hasVersion -->
-      <property name="suppressLoadErrors" value="false"/>
+<!--      <property name="ignoreMethodNamesRegex" value="^get.*$"/>--> <!--It would be nice if we could allow checking that the doc exists without also verifying the @returns is there also, but checkstyle does not allow that. -->
```
[the remark this answers is not in the document]
Comment (tkaitchuck) [reply checkstyle/checkstyle.xml:195]: Done.

**Review by anishakj — APPROVED** [review]
LGTM

**Review by ShwethaSNayak — APPROVED** [review]
LGTM, Looks like code coverage check is failing.

**Review by a6dulaleem — APPROVED** [review]
LGTM, 
only thing is Because of some white spaces added code coverage is failing and I don't know if it propagates to the future PRs.

**Pull request merged**